libcasr/src/lua.rs

```diff
@@ -0,0 +1,337 @@
+        Some(LuaException {
+            message: mat.as_str().to_string(),
+        })
+    }
+    /// Extract stack trace from lua exception.
+    pub fn extract_stacktrace(&self) -> Result<Vec<String>> {
+        LuaStacktrace::extract_stacktrace(&self.message)
+    }
+    /// Transform lua exception into `Stacktrace` type.
+    pub fn parse_stacktrace(&self) -> Result<Stacktrace> {
+        LuaStacktrace::parse_stacktrace(&self.extract_stacktrace()?)
+    }
+    /// Get lua exception as a vector of lines.
+    pub fn lua_report(&self) -> Vec<String> {
+        self.message
+            .split('\n')
+            .map(|s| s.trim().to_string())
+            .collect()
+    }
+}
+
+/// Structure provides an interface for processing the stack trace.
+pub struct LuaStacktrace;
+
+impl ParseStacktrace for LuaStacktrace {
+    fn extract_stacktrace(stream: &str) -> Result<Vec<String>> {
+        let stacktrace = stream
+            .split('\n')
+            .map(|l| l.to_string())
+            .collect::<Vec<String>>();
+        let Some(first) = stacktrace
+            .iter()
+            .position(|line| line.trim().ends_with("stack traceback:"))
+        else {
+            return Err(Error::Casr(
+                "Couldn't find traceback in lua report".to_string(),
+            ));
+        };
+
+        let re = Regex::new(r#".+:(?:|\d+:) in .+"#).unwrap();
+        Ok(stacktrace[first..]
+            .iter()
+            .map(|s| s.trim().to_string())
+            .filter(|s| re.is_match(s))
+            .collect::<Vec<String>>())
+    }
+
+    fn parse_stacktrace_entry(entry: &str) -> Result<StacktraceEntry> {
+        let mut stentry = StacktraceEntry::default();
+        let re = Regex::new(r#"(.+):(\d+): in (\S+ )(\S+)"#).unwrap();
+        let Some(cap) = re.captures(entry) else {
+            return Err(Error::Casr(format!(
+                "Couldn't parse stacktrace line: {entry}"
+            )));
+        };
+        stentry.debug.file = cap.get(1).unwrap().as_str().to_string();
+        if let Ok(line) = cap.get(2).unwrap().as_str().parse::<u64>() {
+            stentry.debug.line = line;
+        } else {
+            return Err(Error::Casr(format!(
+                "Couldn't parse stacktrace line number: {entry}"
+            )));
+        };
+        stentry.function = if let Some(func) = cap.get(4) {
+            func.as_str()
+                .to_string()
+                .trim_matches('\'')
+                .trim_start_matches('<')
+                .trim_end_matches('>')
+                .to_string()
+        } else {
+            return Err(Error::Casr(format!(
+                "Couldn't parse stacktrace function: {entry}"
+            )));
+        };
+        if let Some(appendix) = cap.get(3) {
+            let appendix = appendix.as_str().to_string();
+            if !appendix.starts_with("function")
+                && !appendix.starts_with("upvalue")
+                && !appendix.starts_with("local")
```

Comment:
What else could be here instead of these 3 values? (I saw only `main chunk` in the test)

Reply:
Reworked

Reply:
So, only `main` could be here?

Reply:
Yep, I think we need to remove sufficies before function names, but `main chunk` is an exception